File: LeetCode/0852_PeakIndexInMountainArray/lc0852.cpp
#include <vector>
#include <cassert>

class Solution {
public:
  int peakIndexInMountainArray(std::vector<int>& arr) {
    int left = 0;
    int right = arr.size() - 1;
    while (left < right) {
      int mid = left + (right - left) / 2;
      if (arr[mid] > arr[mid+1]) {
        // current window is in the DESCENDING part of a mountain.
        right = mid;
      }
      else {
        // current window is in the ASCENDING part of a mountain.
        left = mid + 1;
      }
    }
    return left;
  }
};

int main() {
  Solution s{};
  std::vector<int> test1 = {0, 1, 0};
  assert(s.peakIndexInMountainArray(test1) == 1);
  std::vector<int> test2 = {0, 2, 1, 0};
  assert(s.peakIndexInMountainArray(test2) == 1);
  std::vector<int> test3 = {0, 10, 5, 2};
  assert(s.peakIndexInMountainArray(test3) == 1);
}

File: LeetCode/0852/python/lc0852.py
from typing import List

# Runtime complexity: O(log(n))
# Auxiliary space complexity: O(1)
# Subjective level: easy+
# Solved on: 2026-03-10
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        L = len(arr)
        left = 0
        right = L-1
        while left < right:
            mid = (left + right) // 2
            if arr[mid] > arr[mid+1]:
                # mid-to-right contains the descending slope of the mountain.
                # the peak must be somewhere in the left-to-mid part.
                right = mid
            else: # otherwise, arr[mid] < arr[mid+1]. Thus the peak
                # must be somewhere in the mid-to-right part.
                left = mid+1
        return left
